### PR TITLE
Add configurable operation listener for translog

### PR DESCRIPTION
**Pull request by Tim-Brooks**
This commit adds a listener that will be called when new operations are written to the translog writer.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/translog/OperationListener.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+ */
+
+package org.elasticsearch.index.translog;
+
+import org.elasticsearch.common.bytes.BytesReference;
+
+@FunctionalInterface
+public interface OperationListener {
+
+    void operationAdded(BytesReference data, long seqNo, Translog.Location location);
```
[the remark this answers is not in the document]
Comment (Tim-Brooks) [reply server/src/main/java/org/elasticsearch/index/translog/OperationListener.java:16]: I mean I'm not sure we should use an `ActionListener` as we do not have a failure condition here. I understand that we could use a `Consumer<CustomRecord>` but in that scenario we are still adding an abstraction (recover vs. functional interface). So I guess I don't see the advantage to it.

**Review by DaveCTurner — COMMENTED** [review]
Looks ok to me, but could we add some tests that show the listener is called in all the places we expect it to be called?

File: server/src/test/java/org/elasticsearch/index/translog/TranslogTests.java [hunk 2]
```diff
@@ -1579,6 +1597,64 @@ ChannelFactory getChannelFactory() {
...
+        translog.rollGeneration();
+
+        Location location3 = translog.add(TranslogOperationsUtils.indexOp(randomAlphaOfLength(10), 3, primaryTerm.get()));
+        Location location4 = translog.add(TranslogOperationsUtils.indexOp(randomAlphaOfLength(10), 2, primaryTerm.get()));
+        long secondGeneration = translog.getGeneration().translogFileGeneration;
+        assertThat(location3.generation, equalTo(secondGeneration));
+        assertThat(location4.generation, equalTo(secondGeneration));
+
+        try {
+            assertThat(seqNos.get(0), equalTo(0L));
```
Comment (arteam) [inline server/src/test/java/org/elasticsearch/index/translog/TranslogTests.java:1638]: I think we could compare the collections as a whole without doing element by element
```java
assertThat(seqNos, equalTo(List.of(0L, 1L, 3L, 2L)));
assertThat(locations, equalTo(List.of(location1, location2, location3, location4)));
```

**Review by arteam — APPROVED** [review]
LGTM! I've left a few small comments

File: server/src/test/java/org/elasticsearch/index/translog/TranslogTests.java [hunk 3]
```diff
@@ -1579,6 +1597,64 @@ ChannelFactory getChannelFactory() {
...
+        Path tempDir = createTempDir();
+        final Settings settings = Settings.builder().put(IndexMetadata.SETTING_VERSION_CREATED, org.elasticsearch.Version.CURRENT).build();
+
+        final ArrayList<Long> seqNos = new ArrayList<>();
+        final ArrayList<Location> locations = new ArrayList<>();
+        final ArrayList<BytesReference> datas = new ArrayList<>();
+        OperationListener listener = (data, seqNo, location) -> {
+            seqNos.add(seqNo);
+            locations.add(location);
+            try (BytesStreamOutput output = new BytesStreamOutput()) {
```
[the remark this answers is not in the document]
Comment (Tim-Brooks) [reply server/src/test/java/org/elasticsearch/index/translog/TranslogTests.java:1599]: The `data` in the listener is a `ReleasableBytesReference` from the `Translog` that will be released once the method returns. So yes. Obviously in tests it does not matter much, but it does matter for real usages.

I added documentation around the listener for this point.

**Pull request merged**